lib/utils/device_transfer/transfer_protocol.dart

```diff
@@ -18,68 +23,131 @@ const kTypeCommand = 1;
 const kTypeJson = 2;
 const kTypeFile = 3;
 
-/// -----------------------------------------------------------------
-/// | type (1 byte) | body_length（4 bytes） | body | crc（8 bytes） |
+///
 /// ----------------------------------------------------------------
-abstract class TransferPacket {
-  int get _type;
-
-  Future<int> get _bodyLength;
-
-  /// return: body check sum
-  @protected
-  Future<int> _writeBodyToSink(TransferSocket sink);
+/// |                         TransferPacket                      |
+/// ----------------------------------------------------------------
+/// | type(1 byte) | body_length(4 bytes) | body | hMac(32 bytes)|
+/// ----------------------------------------------------------------
+///
+///
+/// ------------------------------------------------------------------------
+/// |                             body                                     |
+/// ------------------------------------------------------------------------
+/// | uuid(16 bytes) for [kTypeFile] | iv(16 bytes) | data (aes encrypted) |
+/// ------------------------------------------------------------------------
+///
+sealed class TransferPacket {
+  TransferPacket();
+
+  Future<void> write(
+    TransferSocket sink, {
+    required Uint8List aesKey,
+    required Uint8List hMacKey,
+  });
 }
 
-abstract class _TransferJsonPacket extends TransferPacket {
-  _TransferJsonPacket(Map<String, dynamic> json)
-      : _data = Uint8List.fromList(utf8.encode(jsonEncode(json)));
-
-  _TransferJsonPacket._fromData(this._data);
-
-  final Uint8List _data;
-
-  @override
-  int get _type => kTypeJson;
-
-  @override
-  Future<int> get _bodyLength => Future.value(_data.length);
-
-  @override
-  Future<int> _writeBodyToSink(TransferSocket sink) async {
-    sink.add(_data);
-    return calculateCrc32(_data);
-  }
+const _kIVBytesCount = 16;
+
+@visibleForTesting
+PaddedBlockCipherImpl createAESCipher({
+  required Uint8List aesKey,
+  required Uint8List iv,
+  required bool encrypt,
+}) {
+  final cbcCipher = CBCBlockCipher(AESEngine());
+  return PaddedBlockCipherImpl(PKCS7Padding(), cbcCipher)
```

Comment:
Please check it. @crossle Is it the same as `PKCS5Padding`?

Reply:
I found out [PKCS#5 padding is a subset of PKCS#7](https://crypto.stackexchange.com/questions/9043/what-is-the-difference-between-pkcs5-padding-and-pkcs7-padding)